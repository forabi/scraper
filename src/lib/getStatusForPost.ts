import {
  groupBy,
  keyBy,
  mapValues,
  omitBy,
  isEmpty,
  mapKeys,
  find,
} from 'lodash';

type Post = {
  id: number;
  post_name: string;
  post_parent: number;
  post_date_gmt: string;
  term_taxonomy_id: number | null;
};

type Term = {
  term_id: number;
  name: string;
  slug: string;
};

type TermTaxonomyRelationship = {
  term_id: number;
  term_taxonomy_id: number;
};

const progressStatus = [
  'pitch',
  'complete',
  'partial',
  'pending-review',
  'postponed',
  'low-priority',
];

export async function getStatusForPost(
  posts: Post[],
  terms: Term[],
  termTaxonomies: TermTaxonomyRelationship[],
) {
  const termIdsByTaxonomyId = keyBy(termTaxonomies, t => t.term_taxonomy_id);

  // Group revisions with their parent post
  const postsByParentId = groupBy(posts, p => {
    if (p.post_parent === 0) {
      return p.id;
    }

    return p.post_parent;
  });

  // An object of the shape `{ 202: 'pitch', 203: 'complete' }`
  const termIdToTermSlug = mapValues(keyBy(terms, t => t.term_id), t => t.slug);

  const termsByParentPostName = mapValues(postsByParentId, v =>
    v
      // Look up the term ID that corresponds to the post's term taxonomy ID
      .map(p => ({
        ...p,
        termId:
          p.term_taxonomy_id !== null && termIdsByTaxonomyId[p.term_taxonomy_id]
            ? termIdsByTaxonomyId[p.term_taxonomy_id].term_id
            : undefined,
      }))
      // Add the term slug ("partial", "low-priority") that corresponds to the the term ID above
      .map(({ termId, ...rest }) => ({
        ...rest,
        termId,
        termSlug: termId !== undefined ? termIdToTermSlug[termId] : undefined,
      }))
      // We are only interested in tags that indicate the status of the post ("partial", "complete"...)
      .filter(
        ({ termSlug }) =>
          termSlug !== undefined && progressStatus.includes(termSlug),
      ),
  );

  const postsById = keyBy(posts, p => p.id);

  // Replace object keys, which are post IDs, with the corresponding post names
  return mapKeys(
    // Remove posts that have no relevant status term
    omitBy(
      // Map each post to it's parent's status
      mapValues(termsByParentPostName, revisions => {
        const revision =
          find(revisions, c => c.post_parent === 0) || revisions[0];

        return revision ? revision.termSlug : undefined;
      }),
      isEmpty,
    ),
    (_, id) => postsById[id].post_name,
  );
}
