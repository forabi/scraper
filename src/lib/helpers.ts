import * as fs from 'fs';
import * as got from 'got';
import { promisify } from 'util';
import * as _glob from 'glob';

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readDir = promisify(fs.readdir);

export const glob = async (pattern: string, options: _glob.IOptions = {}) => {
  return new Promise<string[]>((resolve, reject) => {
    _glob(pattern, options, (error, result) => {
      if (error) {
        reject(error);
      } else {
        resolve(result);
      }
    });
  });
};

export const readJsonFile = async <T = any>(file: string) =>
  readFile(file, 'utf8').then(JSON.parse) as Promise<T>;

export function replaceSmartQuotes(str: string) {
  // prettier-ignore
  return str.replace(/[‘’]/g, '\'').replace(/[“”]/g, '"');
}

export async function fetchPageAsHtml(url: string) {
  const response = await got(url, {
    headers: {
      Accept: 'text/html',
    },
  });

  return response.body;
}
