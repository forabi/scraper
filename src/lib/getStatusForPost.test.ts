import { getStatusForPost } from './getStatusForPost';
import * as path from 'path';

import { readJsonFile } from './helpers';

const fixtures = ['terms.json', 'termTaxonomy.json']
  .map(file => path.join('fixtures', file))
  .map(readJsonFile);

const postsWithChildren = ['tomHanks.json', 'oliviaWilde.json']
  .map(file => path.join('fixtures', 'postsWithChildren', file))
  .map(readJsonFile);

test('gets the correct status for each parent post', async () => {
  const [terms, termTaxonomy] = await Promise.all(fixtures);

  (await Promise.all(postsWithChildren)).forEach(async postWithChildren => {
    expect(
      await getStatusForPost(postWithChildren, terms, termTaxonomy),
    ).toMatchSnapshot();
  });
});
