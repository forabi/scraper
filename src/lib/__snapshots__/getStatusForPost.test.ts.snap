// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets the correct status for each parent post 1`] = `
Object {
  "tom-hanks": "complete",
}
`;

exports[`gets the correct status for each parent post 2`] = `
Object {
  "olivia-wilde": "low-priority",
}
`;
